gpflow/optimizers/scipy.py

```diff
@@ -121,8 +125,13 @@ def callback_func(
 
         def _callback(x: np.ndarray) -> None:
             nonlocal step
-            values = cls.unpack_tensors(variables, x)
-            step_callback(step, variables, values)
+
+            if isinstance(step_callback, Monitor):
+                step_callback(step)
+            else:
+                values = cls.unpack_tensors(variables, x)
+                step_callback(step, variables, values)
```

Comment:
`step_callback` could be a function of one argument without being a `Monitor`, for example, for
```
def foo(x):
    return None
```
`isinstance(f, Monitor)` is False, but `f(step, variables, values)` will still fail. Does this answer the question? I don't think there is any (sensible) way to check at runtime for the number of parameters to a function, but if you only expect a `Monitor` or a 3-parameter function, you could just
```
StepCallback = Union[
    Callable[[int, Sequence[tf.Variable], Sequence[tf.Tensor]], None], Monitor
]
```
Have I understood correctly?

Reply:
Yes, you have! That makes sense... thanks!

Reply:
that said, would it not just be possible to wrap a monitor in a function that ignores some of it arguments, like
```
monitor = Monitor()

def callback(step, variables, values):
    return monitor(step)

optimizer.callback_func(..., callback)
```
or do you particularly care about avoiding the call to `unpack_tensors`

Reply:
Do you mean something similar to https://github.com/GPflow/GPflow/pull/1558? I discussed with STJohn offline - this would cause users to change their code (or use a new API) when they used, for example, Adam as opposed to BFGS/L-BFGS-B...

With the given implementation, these details are hidden from the user?

Or, is there something we can do in the middle which doesn't change the API but also wraps the monitor?

Reply:
ok, if there's a larger reason for it, sure

Reply:
Cool, will keep it as-is then. Thanks for your comments!